core/lib/zksync_core/src/metadata_calculator/pruning.rs

```diff
@@ -0,0 +1,241 @@
+//! Merkle tree pruning logic.
+
+use std::time::Duration;
+
+use anyhow::Context as _;
+use tokio::sync::{oneshot, watch};
+use zksync_dal::{ConnectionPool, Core, CoreDal};
+use zksync_merkle_tree::{MerkleTreePruner, MerkleTreePrunerHandle, RocksDBWrapper};
+
+pub(super) type PruningHandles = (MerkleTreePruner<RocksDBWrapper>, MerkleTreePrunerHandle);
+
+/// Task performing Merkle tree pruning according to the pruning entries in Postgres.
+#[derive(Debug)]
+#[must_use = "Task should `run()` in a managed Tokio task"]
+pub struct MerkleTreePruningTask {
+    handles: oneshot::Receiver<PruningHandles>,
+    pool: ConnectionPool<Core>,
+    poll_interval: Duration,
+}
+
+impl MerkleTreePruningTask {
+    pub(super) fn new(
+        handles: oneshot::Receiver<PruningHandles>,
+        pool: ConnectionPool<Core>,
+        poll_interval: Duration,
+    ) -> Self {
+        Self {
+            handles,
+            pool,
+            poll_interval,
+        }
+    }
+
+    pub async fn run(self, mut stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let (mut pruner, pruner_handle);
+        tokio::select! {
+            res = self.handles => {
+                match res {
+                    Ok(handles) => (pruner, pruner_handle) = handles,
+                    Err(_) => {
+                        tracing::info!("Merkle tree dropped; shutting down tree pruning");
+                        return Ok(());
+                    }
+                }
+            }
+            _ = stop_receiver.changed() => {
+                tracing::info!("Stop signal received before Merkle tree is initialized; shutting down tree pruning");
+                return Ok(());
+            }
+        }
+        tracing::info!("Obtained pruning handles; starting Merkle tree pruning");
+
+        // FIXME: is this good enough (vs a managed task)?
+        pruner.set_poll_interval(self.poll_interval);
+        let pruner_task_handle = tokio::task::spawn_blocking(|| pruner.run());
```

Comment:
Subj. On one hand, `pruner` can panic, which won't be detected immediately by the current setup (it could be changed to do that; e.g., by checking the reference count of `target_retained_version` in `pruner_handle.set_target_retained_version` and returning an error if it's 1). On the other hand, since it's a blocking task, it cannot be stopped and fundamentally depends on `MerkleTreePruningTask` to be cancel-aware.